core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java

```diff
@@ -141,33 +138,49 @@ private List<Result> scanInternal(Scan originalScan) throws CrudException {
 
     Optional<Map<Snapshot.Key, TransactionResult>> resultsInSnapshot = snapshot.getResults(scan);
     if (resultsInSnapshot.isPresent()) {
-      return createScanResults(scan, originalProjections, resultsInSnapshot.get());
+      return createFilteredScanResults(scan, originalProjections, resultsInSnapshot.get());
     }
 
+    // Results for users
     Map<Snapshot.Key, TransactionResult> results = new LinkedHashMap<>();
 
+    // Results for the scan set
+    Map<Snapshot.Key, TransactionResult> actualResults = new LinkedHashMap<>();
+
     Scanner scanner = null;
     try {
-      scanner = scanFromStorage(scan);
-      for (Result r : scanner) {
-        TransactionResult result = new TransactionResult(r);
-        if (!result.isCommitted()) {
-          throw new UncommittedRecordException(
-              scan,
-              result,
-              CoreError.CONSENSUS_COMMIT_READ_UNCOMMITTED_RECORD.buildMessage(),
-              snapshot.getId());
-        }
+      @Nullable Integer limit = null;
+      if (scan.getLimit() == 0) {
+        scanner = scanFromStorage(scan);
+      } else {
+        limit = scan.getLimit();
+
+        // Get a scanner without the limit if the scan has a limit
+        scanner = scanFromStorage(Scan.newBuilder(scan).limit(0).build());
```

Comment:
This looks good. But, it seems to fetch all records. So, how about this optimization?
- At this line, set the limit to the original limit plus the size of `snapshot.deleteSet` considering cases where the scan range overlaps with `snapshot.deleteSet`
- Skip a result if it's contained in `snapshot.deleteSet`

Reply:
Yeah, I'm not sure all the storages will be affected or not. But, I guess either of taking shared locks or equivalents, loading records in memory as a read-set, updating read-timestamps, or etc would internally occur in the underlying database.

I also think users want to use LIMIT when the number of records associated with a partition key will be large. So, pushing down the limit condition (or a bit modified one) as much as possible should be useful.

Reply:
Understood. Since the current code doesn’t fetch all records at once and retrieves them one by one through the `Scanner` instance, I don’t think it causes any significant performance issues at this point. I’d prefer to keep the code simple for now and address this as a separate optimization task later if any performance issues arise. Thanks!

Reply:
> I’d prefer to keep the code simple for now

I guess the reason why `validateScanResults()` uses the double iterators is related to this change to use no-limit. If so, honestly speaking, the implementation using the double iterators doesn't seem simple to me, compared to the original implementation.

Reply:
Let's discuss it in another PR. Thanks!